### DISCUSSION TITLE
Hitting Datadog rate limits (EKS); how to query `kube-state-metrics` values in-cluster?

### DISCUSSION THREAD

**Discussion by kencieszykowski** [kedacore/keda#3909] [Q&A / Need Help] [answered]
Hey everyone!  

We had previously been using the Datadog Cluster Agent to register the External Metrics API; but when we discovered KEDA we thought that all of the additions would be huge.

Here is everything deployed (apologies, don't have version numbers at this point; but hopefully that's irrelevant at this point):
- Datadog Agent + Cluster Agent via Helm Chart
-- kube-state-metrics is deployed as a dependency with Datadog as well
- kubernetes-prometheus-stack deployed [via Helm chart](https://github.com/prometheus-community/helm-charts/tree/kube-prometheus-stack-14.5.0/charts/kube-prometheus-stack#multiple-releases), with dependency for KSM turned off (if the scraper is turned on, this will send out _another_ deployment of KSM
- `scaledObject` that's simply monitoring the Readiness of pods (cannot use Kubernetes Workload scaler since the target deployment exists in another namespace)

The polling interval for Datadog metrics is locked in at 15s, as we don't have control over the HPA interval on `kube-controller-manager`, since we're using EKS.  I've reached out to Datadog support to increase this limit-- but if we're hitting it after only _four_ scaledObjects are using it, I'm not convinced that this will be a long term solution.

So I'm just trying to figure out how to wade through these things and instead get `kube-state-metrics` values scaped via Prometheus, and then use that Scaler.  It _seems_ like I should just pull KSM off of the Datadog Helm chart, and instead install it alongside Prometheus.  This will enable the Prometheus scraper for the dependent deployment, and since KSM is available on a NodePort, the Datadog integration should be able to query this as expected and pass the details into Datadog as well.  

Please let me know if I'm overlooking an easy solution here, or if I'm completely thinking of this issue the wrong way.  Appreciate the help!

**Comment by tomkerkhove** [comment]
I'm not familiar with the Datadog Cluster Agent but is there a reason why you don't simply standardize on the Datadog scaler or am I missing something?

**Reply by kencieszykowski** [reply]
@tomkerkhove 

Apologies if I wasn't clear there-- we did standardize on the Datadog Scaler.  I just added the Cluster Agent stuff to give some background that we were pulling external metrics through it previously without hitting a rate limit; which leads me to believe that signed code directly from DD either works in a different way entirely, or effectively gets a pass for rate limiting.

**Comment by kencieszykowski** [comment]
For more information: I currently have a ticket in with Datadog to increase our rate limit on the API endpoint ( I believe KEDA is using the `Query timeseries points` API) ; but just trying to find some other solutions in the case this doesn't go through.

One thing I was curious about (that may warrant its' own discussion post); is that we have several different applications keying off of the same Datadog metric query; but it seems like all of these get sent as separate requests.  Curious on how Keda handles this in the back end, and if checking for an existing query is possible before registering a new metric on the server.

**Comment by kencieszykowski** [comment, marked as answer]
Mmk; scratch that-- ended up tossing a Prometheus instance that scrapes `datadog-kube-state-metrics` in the Cluster.  Prometheus scaler is working well and will supply us with the information we need that isn't rate limited.  Closing this one up!
